sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/CharsetDetector.java

```diff
@@ -77,52 +64,27 @@ public InputStream inputStream() {
     return stream;
   }
 
-  private static BOMInputStream createInputStream(Path path) throws IOException {
-    BufferedInputStream bufferedStream = new BufferedInputStream(Files.newInputStream(path));
-    return new BOMInputStream(bufferedStream, ByteOrderMark.UTF_8, ByteOrderMark.UTF_16LE,
-      ByteOrderMark.UTF_16BE, ByteOrderMark.UTF_32LE, ByteOrderMark.UTF_32BE);
-  }
-
-  private boolean detectBOM() throws IOException {
-    String charsetName = stream.getBOMCharsetName();
-    if (charsetName != null) {
-      detectedCharset = Charset.forName(charsetName);
-      return true;
-    }
-    return false;
-  }
-
-  private boolean detectCharset() throws IOException {
+  private byte[] readBuffer() throws IOException {
+    stream = new BufferedInputStream(Files.newInputStream(filePath), 8192);
     stream.mark(BYTES_TO_DECODE);
     byte[] buf = new byte[BYTES_TO_DECODE];
-    int len = IOUtils.read(stream, buf, 0, BYTES_TO_DECODE);
+    int read = IOUtils.read(stream, buf, 0, BYTES_TO_DECODE);
     stream.reset();
-
-    Set<Charset> charsets = new LinkedHashSet<>();
-    charsets.add(defaultEncoding);
-    charsets.add(StandardCharsets.UTF_8);
-    charsets.add(Charset.defaultCharset());
-
-    for (Charset c : charsets) {
-      if (tryDecode(buf, len, c)) {
-        detectedCharset = c;
-        return true;
-      }
-    }
-    return false;
+    stream.mark(-1);
+    return Arrays.copyOf(buf, read);
   }
 
-  private static boolean tryDecode(byte[] bytes, int len, Charset charset) throws IOException {
-    CharsetDecoder decoder = charset.newDecoder()
-      .onMalformedInput(CodingErrorAction.REPORT)
-      .onUnmappableCharacter(CodingErrorAction.REPORT);
-
-    try {
-      decoder.decode(ByteBuffer.wrap(bytes, 0, len));
-    } catch (CharacterCodingException e) {
-      return false;
+  private boolean detectCharset(byte[] buf) throws IOException {
+    ByteCharsetDetector detector = new ByteCharsetDetector(new CharsetValidation(), userEncoding);
+    ByteOrderMark bom = detector.detectBOM(buf);
+    if (bom != null) {
+      detectedCharset = Charset.forName(bom.getCharsetName());
+      stream.skip(bom.length());
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Check the return value of the "skip" call to see how many bytes were read. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2674)